src/java/org/apache/cassandra/config/DataStorage.java

```diff
@@ -0,0 +1,378 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.cassandra.config;
+
+import java.util.Arrays;
+import java.util.Objects;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import com.google.common.primitives.Ints;
+
+/**
+ * Represents an amount of data storage
+ */
+public final class DataStorage
+{
+    /**
+     * The Regexp used to parse the storage provided as String.
+     */
+    private static final Pattern STORAGE_UNITS_PATTERN = Pattern.compile("^(\\d+)(kb|KB|mb|MB|gb|GB|b|B)$");
+
+    private final long quantity;
+
+    private final DataStorageUnit unit;
+
+    public DataStorage(String value)
+    {
+        if(value == null || value.equals("null"))
+        {
+            quantity = 0;
+            unit = DataStorageUnit.MEGABYTES; // the unit doesn't really matter as 0 is 0 in all units
+            return;
+        }
+        //parse the string field value
+        Matcher matcher = STORAGE_UNITS_PATTERN.matcher(value);
+
+        if (!matcher.find())
+        {
+            throw new IllegalArgumentException("Invalid data storage size: " + value);
+        }
+
+        quantity = Long.parseLong(matcher.group(1));
+        unit = DataStorageUnit.fromSymbol(matcher.group(2));
+    }
+
+    private DataStorage(long quantity, DataStorageUnit unit)
+    {
+        this.quantity = quantity;
+        this.unit = unit;
+    }
+
+    /**
+     * Creates a {@code DataStorage} of the specified amount of bytes.
+     *
+     * @param bytes the amount of bytes
+     * @return a {@code DataStorage}
+     */
+    public static DataStorage inBytes(long bytes)
```

Comment:
The way I read it is we create a DataStorage in Bytes, we won't take it from bytes to create object in megabytes, it will be object in Bytes. I feel fromBytes can be misread. We create a DataStorage from bytes in bytes. 
We don't stress on from but on the unit that DataStorage object created is in.